Persisted config source not loading properties at runtime

pedroigor opened the pull request:
Closes #12156

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

pedroigor committed:
Persisted config source not loading properties at runtime

Closes #12156

Co-authored-by: Dominik Guhr <dguhr@redhat.com>

pedroigor commented:
Looks like nexus is unavailable.

DGuhr reviewed (changes requested):
@pedroigor the fix itself in the codebase LGTM, I wasn't sure if checking for the ClassLoader instance was right at all, wanted to discuss with you. Seems not :) 

For the testing enhancements:
- When no additional test dependencies should be added (as per [this](https://github.com/keycloak/keycloak/pull/12081#issuecomment-1135235596) comment), why you also add 2? Easily achievable using only the Testcontainers library we have, and `GenericContainer` with the appropriate images ([ref](https://www.testcontainers.org/features/creating_container/)), but a bit more boilerplate involved (setting DB user, pw etc using `withEnv` or sth), so I thought add the dependency, only for test for sure, to avoid this boilerplate. Feel free to leave them, as it is what I did essentially, it's just that I don't get that comment then ;)
- pls incorporate my tests in this PR in order to make it approvable. Looking at your code, you set `db` using `dist.setProperty()` so it would be available from non-persistent source in the properties file we generate unless we have a `removeProperty` that we don't have or we use `ReInstall.BEFORE_TEST` that makes no sense. 
To make it more concise: [This](https://github.com/keycloak/keycloak/pull/12081/files#diff-02217218d23673cca59a7392aed483d6c229ee8568668aeafee58e217c0ef326R56-R59) test that I introduced to make sure the `db` prop is read from persisted properties would fail with your current PR I guess, bc. `db` is actually read from the properties file in the temp kc installation instead of being read from the persisted properties. This would lead to the subsequent tests actually succeeding (see proposed workaround [here](https://github.com/keycloak/keycloak/issues/11902#issuecomment-1127553685)), but is not the use case we want to test => using a pre-optimized serverimage _without_ the `db` set anywhere but in persisted props.
- w.r.t. "adding functionality for additional databases": I avoided this to not add unnecessary complexity until it is needed. But when it works, feel free to add it for mariadb, too, but then there's the question why not add it for all DBs? Imo this is a point for an extra ticket or should be incorporated for additional DBs whenever it is needed by a specific use case. My PR introduced "just enough" code to fix the issue. I used postgres bc. it is the only db the operator also supports, and one of the two dbs we want to go with later on as first-class, so imo this should be our go-to db for testing. Also, the problem was not really db-specific, but configuration and build/runtime behaviour specific so one test for one db should suffice w.r.t. test execution times. That said, _when_ it works, why not..

*hushs back to PTO ;)

pedroigor commented:
> @pedroigor the fix itself in the codebase LGTM, I wasn't sure if checking for the ClassLoader instance was right at all, wanted to discuss with you. Seems not :)
> 
> For the testing enhancements:
> 
> * When no additional test dependencies should be added (as per [this](https://github.com/keycloak/keycloak/pull/12081#issuecomment-1135235596) comment), why you also add 2? Easily achievable using only the Testcontainers library we have, and `GenericContainer` with the appropriate images ([ref](https://www.testcontainers.org/features/creating_container/)), but a bit more boilerplate involved (setting DB user, pw etc using `withEnv` or sth), so I thought add the dependency, only for test for sure, to avoid this boilerplate. Feel free to leave them, as it is what I did essentially, it's just that I don't get that comment then ;)

Sorry, I meant you don't need to set a version if you have a transient dependency. The way you did will force us to change versions between Quarkus upgrades.

Keep in mind that we should leverage dependencies (and versions) from Quarkus as much as possible

> * pls incorporate my tests in this PR in order to make it approvable. Looking at your code, you set `db` using `dist.setProperty()` so it would be available from non-persistent source in the properties file we generate unless we have a `removeProperty` that we don't have or we use `ReInstall.BEFORE_TEST` that makes no sense.
>   To make it more concise: [This](https://github.com/keycloak/keycloak/pull/12081/files#diff-02217218d23673cca59a7392aed483d6c229ee8568668aeafee58e217c0ef326R56-R59) test that I introduced to make sure the `db` prop is read from persisted properties would fail with your current PR I guess, bc. `db` is actually read from the properties file in the temp kc installation instead of being read from the persisted properties. This would lead to the subsequent tests actually succeeding (see proposed workaround [here](https://github.com/keycloak/keycloak/issues/11902#issuecomment-1127553685)), but is not the use case we want to test => using a pre-optimized serverimage _without_ the `db` set anywhere but in persisted props.

I've added the test to check `show-config` and run all tests without the `db` option set in the configuration file.

> * w.r.t. "adding functionality for additional databases": I avoided this to not add unnecessary complexity until it is needed. But when it works, feel free to add it for mariadb, too, but then there's the question why not add it for all DBs? Imo this is a point for an extra ticket or should be incorporated for additional DBs whenever it is needed by a specific use case. My PR introduced "just enough" code to fix the issue. I used postgres bc. it is the only db the operator also supports, and one of the two dbs we want to go with later on as first-class, so imo this should be our go-to db for testing. Also, the problem was not really db-specific, but configuration and build/runtime behaviour specific so one test for one db should suffice w.r.t. test execution times. That said, _when_ it works, why not..

There is no additional complexity, we are just leveraging the capabilities available from the test suite to also allow bootstrapping a container when running distribution tests. No need to manually create containers within tests but just use the annotation we already have. We should try to enhance the testsuite whenever we have a chance and here we are now able to run tests against the distribution using a database (in a generic way).

> 
> *hushs back to PTO ;)

Let's talk more once you're back. The latest changes should be enough to introduce your test scenario.

andreaTP commented on quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/DatabaseContainer.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    String getUsername() {
+        return "keycloak";
```
bind it to a `static` field ? e.g. `DEFAULT_USERNAME` ?

pedroigor replied:
Why? It should be in the constant pool, doesn't it? Different than the default password, this one is only used within this class so I did not find it necessary to create a constant.

andreaTP commented on quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/WithDatabase.java:
```diff
@@ -33,8 +33,11 @@
+    /**
+     * If the {@code db} option must be unset after the running the build command.
+     */
+    boolean removeBuildOptionsAfterBuild() default false;
```
I believe that this should default to `true`

pedroigor replied:
Why?

pedroigor replied:
Ok

pedroigor replied:
Actually, I would prefer to not do it now because we need to look at other tests and see how this option changes their expectations.

We do have tests covering this area but for other situations and configuration options. It should be fine to keep it `false` and enabled only for this test.

andreaTP commented on quarkus/tests/integration/src/main/java/org/keycloak/it/utils/RawKeycloakDistribution.java:
```diff
@@ -412,6 +412,11 @@ public void setProperty(String key, String value) {
 
+    @Override
+    public void removeProperty(String db) {
+        setProperty(db, null, distPath.resolve("conf").resolve("keycloak.conf").toFile());
```
How the implementation for the `DockerKeycloakDistribution` would look like?

pedroigor replied:
The same as `setProperty` we have already have: unsupported.

andreaTP replied:
👍

andreaTP commented on quarkus/tests/integration/src/main/java/org/keycloak/it/utils/RawKeycloakDistribution.java:
```diff
@@ -456,7 +461,11 @@ private void setProperty(String key, String value, File confFile) {
+            if (value != null) {
+                properties.put(key, value);
+            } else {
+                properties.remove(key);
```
this is unexpected as the method is called `setProperty`, is it needed?

pedroigor replied:
Yes it is needed to allow removing properties from the config if needed. Do you want to change the method name? `setOrRemoveProperty`? Isn't enough to accept that a `null` value invalidates a key?

pedroigor replied:
I've changed to use `removeProperty`.

andreaTP commented on quarkus/tests/integration/src/test/java/org/keycloak/it/storage/database/BasicDatabaseTest.java:
```diff
@@ -43,12 +47,31 @@ void testWrongUsername(LaunchResult result) {
 
     protected abstract void assertWrongPassword(CLIResult cliResult);
+
+    @Order(1)
```
I would prefer to make those tests order-agnostic committing to the repo a minimal export. wdyt?

pedroigor replied:
Could you elaborate, please? This is the same approach from @DGuhr PR that you approved.

andreaTP commented on quarkus/tests/integration/src/test/java/org/keycloak/it/storage/database/BasicDatabaseTest.java:
```diff
@@ -17,23 +17,27 @@
 
     @Test
-    @Launch({ "start-dev" })
+    @Launch({ "start", "--http-enabled=true", "--hostname-strict=false" })
```
Have you considered adding the tests for `start` instead of swapping from `start-dev`?

pedroigor replied:
Could you please elaborate?

pedroigor replied:
Still not sure what you mean ... We have tests for both commands in the distribution. Here we want to check how the `start` behaves with a pre-built image.

andreaTP commented on quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/CLITestExtension.java:
```diff
@@ -237,10 +243,26 @@ private void configureDatabase(ExtensionContext context) {
+
+                dist.start(List.of("build"));
+
+                if (database.removeBuildOptionsAfterBuild()) {
```
Reflecting the encoding `removeBuildOptionsAfterBuild` -> `removeDBOptionAfterBuild`, wdyt?

andreaTP replied:
> it should be generic enough to remove any build option set when configuring the database.

fine for me if that's the behavior, in this case, we are only removing the `db` property, we can use the `PropertyMapper` and iterate all the build time properties.

pedroigor replied:
You make me wonder if this option makes more sense at the `DistributionTest` annotation instead because it is not specific to database tests.

I've updated the PR and moved this option to reflect that. Looks better to me now.

andreaTP reviewed (commented):
Nice work! Added a few comments

andreaTP commented:
Side note: the Operator CI might be fixed by those changes:
https://github.com/keycloak/keycloak/pull/11642

pedroigor committed:
fixup! Persisted config source not loading properties at runtime

andreaTP reviewed (approved):
LGTM
